### PR TITLE
further improvements to thumbnail placement in culling

### PR DESCRIPTION
**Pull request by Solarer**
@ Reviewer: I have not cleaned up / squashed the commits yet because I want to get the functionality of this PR approved first before I do a final cleaning. Since I changed many code lines I will try to organize this in a way to make review easier.

This is a follow up PR to #14966 which left a few issues unaddressed.
* sometimes images do not have the same height (see #12875)
* 'stacking' functionality was completely borked

Fixed issues:
* Stacking now works as intended without overlapping thumbnails or leaving the screen. To reproduce, display 3 or more thumbnails where the combined height of 2 thumbnails is smaller than the third.

original             |  fixed
:-------------------------:|:-------------------------:
image  |  image

* Thumbnails are now scaled to equal height. 
To avoid gigantic thumbnails when scaling very slim thumbnails, I limited this to 2 times the average thumbnail width.
This does even out slight differences (example 1) but also makes better use of the available space when displaying mixed aspect ratios of portrait and landscape images (example 2)

original             |  fixed
:-------------------------:|:-------------------------:
image  |  image
image  |  image

I personally found this irritating when looking at only 2 images (one a landscape and the other a portrait) since the landscape will appear to be bigger than the portrait. Therefore I limited scaling to 20% if only 2 images are displayed. This does even out small differences but portrait and landscape will not be scaled to the same height.

without 20% limit (just a test, I do not like this)            |  with 20% limit (my final implementation - looks like before)
:-------------------------:|:-------------------------:
image  | image

* Scaling all images to the same height allowed me to simplify the thumbnail placement algorithm and achieve better placement results in some situations

original             |  fixed
:-------------------------:|:-------------------------:
image  |  image

### CODE REVIEW

**Comment by TurboGit** [comment]
First test on my side and this looks really better. I was about to create an issue yesterday as on two duplicated images I had not a proper synch of the view when zooming in (that is, one image was zoomed a bit more than the other even-though they were true duplicate - only one was B&W). This is now working properly on my side, so this issue is also fixed here.

All in all, this seems really good to me.

File: src/dtgtk/culling.c [hunk 1]
```diff
@@ -1645,25 +1547,130 @@ static gboolean _thumbs_compute_positions(dt_culling_t *table)
...
     if(!slot_iter)
     {
       slots = g_list_prepend(slots, g_list_prepend(NULL, thumb));
+      max_slot_heigth = MAX(max_slot_heigth, thumb->height);
     }
-
   }
   slots = g_list_reverse(slots);  // list was built in reverse order, so un-reverse it
+  const int number_of_slots = g_list_length(slots);
+
```
Comment (TurboGit) [inline src/dtgtk/culling.c:1564]: duplicate empty line - can be removed

File: src/dtgtk/culling.c [hunk 2]
```diff
@@ -1645,25 +1547,130 @@ static gboolean _thumbs_compute_positions(dt_culling_t *table)
...
+    }
+    scaled_slot_height -= spacing;
+    total_slot_width += scaled_slot_width + spacing;
+
+    // iterative formula to calculate average slot ratio and width
+    avg_slot_aspect_r += (scaled_slot_width/(float)scaled_slot_height - avg_slot_aspect_r) / (float)slot_counter;
+    avg_slot_width += (scaled_slot_width - avg_slot_width) / (float)thumb_counter;
+  }
+  total_slot_width -= spacing;
+
```
Comment (TurboGit) [inline src/dtgtk/culling.c:1630]: likewise

File: src/dtgtk/culling.c [hunk 3]
```diff
@@ -1645,25 +1547,130 @@ static gboolean _thumbs_compute_positions(dt_culling_t *table)
...
+    if(planned_total_width / (float)planned_total_height > screen_aspect_r)
+      row_cnt_tmp = row_cnt + 1;
+    else
+      row_cnt_tmp = row_cnt - 1;
+
+    if(row_cnt_tmp == 0 || row_cnt_tmp > slot_counter)
+      break;
+
+    float planned_total_width_tmp = total_slot_width / (float) row_cnt_tmp;
```
Comment (TurboGit) [inline src/dtgtk/culling.c:1657]: const

File: src/dtgtk/culling.c [hunk 4]
```diff
@@ -1645,25 +1547,130 @@ static gboolean _thumbs_compute_positions(dt_culling_t *table)
...
+    int planned_total_height_tmp = row_cnt_tmp * max_slot_heigth;
```
Comment (TurboGit) [inline src/dtgtk/culling.c:1658]: const

File: src/dtgtk/culling.c [hunk 5]
```diff
@@ -1673,37 +1680,41 @@ static gboolean _thumbs_compute_positions(dt_culling_t *table)
...
-          slot_cw_iter;
-          slot_cw_iter = g_list_next(slot_cw_iter))
+      for(GList *slot_thumb_iter = slot;
+          slot_thumb_iter;
+          slot_thumb_iter = g_list_next(slot_thumb_iter))
       {
-        dt_thumbnail_t *cw = (dt_thumbnail_t *)slot_cw_iter->data;
-        slot_max_thumb_width = MAX(slot_max_thumb_width, cw->width);
-        slot_total_heigth = slot_total_heigth + cw->height + spacing;
+        dt_thumbnail_t *thumb = (dt_thumbnail_t *)slot_thumb_iter->data;
```
Comment (TurboGit) [inline src/dtgtk/culling.c:1687]: const

File: src/dtgtk/culling.c [hunk 6]
```diff
@@ -1721,70 +1732,105 @@ static gboolean _thumbs_compute_positions(dt_culling_t *table)
...
+      int slot_heigth = 0;
 
+      // loop through thumbs of the slot
+      // to calculate slot heigth and update row width and heigth
+      // which is used for xoffset of row and yoffset of individual thumbs
+      for(GList *slot_thumb_iter = slot;
+        slot_thumb_iter;
+        slot_thumb_iter = g_list_next(slot_thumb_iter))
+      {
+        dt_thumbnail_t *thumb = (dt_thumbnail_t *)slot_thumb_iter->data;
```
Comment (TurboGit) [inline src/dtgtk/culling.c:1786]: const

**Review by TurboGit — CHANGES_REQUESTED** [review]
Some minor style issues.

**Comment by Solarer** [comment]
@TurboGit I fixed the style issues.
Did you check for them manually or did you run a command that showed you missing `const` and duplicated newlines?

**Comment by TurboGit** [comment]
> Did you check for them manually or did you run a command that showed you missing `const` and duplicated newlines?

By code reading.

**Review by TurboGit — APPROVED** [review]
Very nice improvements. Thanks!

**Pull request merged**